drv/lpc55-swd/src/main.rs

```diff
@@ -1748,195 +1725,151 @@ impl ServerImpl {
-            ringbuf_entry!(Trace::IncompleteUndo(need_undo));
-            error = true;
+    /// Disables vector catch by clearing `DEMCR.VC_CORERESET`
+    fn disable_vector_catch(&mut self) {
+        if self
+            .dp_write_bitflags::<Demcr>(Demcr::from_bits_retain(0))
+            .is_err()
+        {
+            ringbuf_entry!(Trace::DemcrWriteError);
         }
+    }
 
-        // The SP is still halted.
-        // Get it running again by toggling its RESET pin.
-        self.sp_reset_enter();
-
-        // Record a successful measurement before releasing the SP from reset.
-        let success = if let Ok(digest) = digest {
-            // SP resets the attestation log and record the new measurement.
-            if !error
-                && self
+    /// Resets the SP and obtains a measurement, sending it to `attest`
+    ///
+    /// Returns `Ok(())` if the SP was measured, `Err(())` if something went
+    /// wrong and existing measurements should be invalidated.
+    fn reset_and_measure_sp(&mut self) -> Result<(), ()> {
+        self.reset_into_debug_halt()?;
+        let success = match self.do_measure_sp() {
+            Ok(digest) => {
+                // SP resets the attestation log and record the new measurement.
+                if self
                     .attest
                     .reset_and_record(HashAlgorithm::Sha3_256, &digest)
                     .is_ok()
-            {
-                ringbuf_entry!(Trace::RecordedMeasurement);
-                true
-            } else {
-                ringbuf_entry!(Trace::RecordMeasurementFailed);
-                false
+                {
+                    ringbuf_entry!(Trace::RecordedMeasurement);
+                    Ok(())
+                } else {
+                    ringbuf_entry!(Trace::RecordMeasurementFailed);
+                    Err(())
+                }
+            }
+            Err(()) => {
+                ringbuf_entry!(Trace::MeasureFailed);
+                Err(())
             }
-        } else {
-            ringbuf_entry!(Trace::MeasureFailed);
-            false
         };
 
-        hl::sleep_for(1);
-        self.sp_reset_leave();
-
-        let now = sys_get_timer().now;
-        ringbuf_entry!(Trace::MeasuredSp {
-            success,
-            delta_t: (now.saturating_sub(start)) as u32
-        });
-
+        if success.is_ok() && self.reset_into_debug_halt().is_ok() {
+            // Deposit the measurement token into SP memory
+            if let Err(e) = self.write_single_target_addr(
+                measurement_token::MEASUREMENT_BASE as u32,
+                measurement_token::MEASUREMENT_TOKEN_VALID,
+            ) {
+                ringbuf_entry!(Trace::TokenWriteFail(e));
+            }
+            self.disable_halting_debug();
+        } else {
+            // Reset the SP into normal operation
+            self.disable_halting_debug();
+            self.sp_reset_enter();
+            hl::sleep_for(1);
```

Comment:
this is a place where i might want to see a comment about "where does 1ms come from", but maybe there isn't an obvious "it says at least this long in <PLACE>" for it?